pkg/function/restore_csi_snapshot.go

```diff
@@ -0,0 +1,195 @@
+	// RestoreCSISnapshotStorageRequestArg provides the storage size to be requested for PV in PVC
+	RestoreCSISnapshotRestoreSizeArg = "restoreSize"
+	// RestoreCSISnapshotAccessModesArg lists down the accessmodes for the underlying PV
+	RestoreCSISnapshotAccessModesArg = "accessModes"
+	// RestoreCSISnapshotLabelsArg has labels that will be added to the newly restored PVC
+	RestoreCSISnapshotLabelsArg = "labels"
+	// RestoreCSISnapshotVolumeModeArg defines mode of volume
+	RestoreCSISnapshotVolumeModeArg = "volumeMode"
+)
+
+type restoreCSISnapshotFunc struct{}
+
+type restoreCSISnapshotArgs struct {
+	Name         string
+	PVC          string
+	Namespace    string
+	StorageClass string
+	RestoreSize  *resource.Quantity
+	AccessModes  []v1.PersistentVolumeAccessMode
+	Labels       map[string]string
+	VolumeMode   v1.PersistentVolumeMode
+}
+
+func (*restoreCSISnapshotFunc) Name() string {
+	return RestoreCSISnapshotFuncName
+}
+
+func (*restoreCSISnapshotFunc) Exec(ctx context.Context, tp param.TemplateParams, args map[string]interface{}) (map[string]interface{}, error) {
+	var restoreSize string
+	var restoreArgs restoreCSISnapshotArgs
+	if err := Arg(args, RestoreCSISnapshotNameArg, &restoreArgs.Name); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, RestoreCSISnapshotPVCNameArg, &restoreArgs.PVC); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, RestoreCSISnapshotNamespaceArg, &restoreArgs.Namespace); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, RestoreCSISnapshotStorageClassArg, &restoreArgs.StorageClass); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, RestoreCSISnapshotRestoreSizeArg, &restoreSize); err != nil {
+		return nil, err
+	}
+	if err := OptArg(args, RestoreCSISnapshotAccessModesArg, &restoreArgs.AccessModes, []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce}); err != nil {
+		return nil, err
+	}
+	if err := validateVolumeAccessModesArg(restoreArgs.AccessModes); err != nil {
+		return nil, err
+	}
+	if err := OptArg(args, RestoreCSISnapshotVolumeModeArg, &restoreArgs.VolumeMode, v1.PersistentVolumeFilesystem); err != nil {
+		return nil, err
+	}
+	if err := validateVolumeModeArg(restoreArgs.VolumeMode); err != nil {
+		return nil, err
+	}
+	if err := OptArg(args, RestoreCSISnapshotLabelsArg, &restoreArgs.Labels, nil); err != nil {
+		return nil, err
+	}
+	size, err := resource.ParseQuantity(restoreSize)
+	if err != nil {
+		return nil, err
+	}
+	restoreArgs.RestoreSize = &size
+
+	if err := restoreCSISnapshot(ctx, restoreArgs); err != nil {
+		return nil, err
+	}
+	return nil, nil
+}
+
+func (*restoreCSISnapshotFunc) RequiredArgs() []string {
+	return []string{
+		RestoreCSISnapshotNameArg,
+		RestoreCSISnapshotPVCNameArg,
+		RestoreCSISnapshotNamespaceArg,
+		RestoreCSISnapshotStorageClassArg,
+		RestoreCSISnapshotRestoreSizeArg,
+	}
+}
+
+func restoreCSISnapshot(ctx context.Context, args restoreCSISnapshotArgs) error {
+	kubeCli, err := kube.NewClient()
+	if err != nil {
+		return err
+	}
```

Comment:
I went through the test again and realized we are not really testing restore but just creating the PVC directly which is what restore is also doing. But to test restore, I would recommend moving this part out to `Exec` and pass in the `cli` to this function. That way, you can directly call `restoreCSISnapshot` in the test with the fake CLI.

Reply:
A similar refactor can be done to the `createCSISnapshot` function as well. I would move the functionality between line 99 to 116 into a helper function that accepts `ctx, snapshotter, args...`. It makes testing easier without having to rewrite all the steps already taken care of by the production code.

Reply:
Addressed both comments. Create snapshot chnages are present in this PR https://github.com/kanisterio/kanister/pull/1186